components/net/http_cache.rs

```diff
@@ -0,0 +1,488 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#![deny(missing_docs)]
+
+//! A memory cache Implements the logic specified in http://tools.ietf.org/html/rfc7234
+//! and http://tools.ietf.org/html/rfc7232.
+
+use fetch::methods::DoneChannel;
+use hyper::header;
+use hyper::header::ContentType;
+use hyper::header::Headers;
+use hyper::method::Method;
+use hyper::status::StatusCode;
+use hyper_serde::Serde;
+use net_traits::{Metadata, FetchMetadata, FilteredMetadata};
+use net_traits::request::Request;
+use net_traits::response::{Response, ResponseBody};
+use servo_url::ServoUrl;
+use std::collections::HashMap;
+use std::str::FromStr;
+use std::sync::{Arc, Mutex};
+use time;
+use time::{Duration, Tm};
+use url::Url;
+
+
+/// The key used to differentiate requests in the cache.
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct CacheKey {
+    url: ServoUrl
+}
+
+impl CacheKey {
+    fn new(request: Request) -> CacheKey {
+        CacheKey {
+            url: request.url().clone()
+        }
+    }
+
+    fn from_url_string(url: String) -> CacheKey {
+        let url = Url::parse(&url).unwrap();
+        CacheKey {
+            url: ServoUrl::from_url(url)
+        }
+    }
+
+    /// Retrieve the URL associated with this key
+    pub fn url(&self) -> ServoUrl {
+        self.url.clone()
+    }
+}
+
+/// A complete cached resource.
+#[derive(Clone, Debug)]
+struct CachedResource {
+    metadata: CachedMetadata,
+    request_headers: Vec<(String, String)>,
+    body: Arc<Mutex<ResponseBody>>,
+    status: Option<StatusCode>,
+    raw_status: Option<(u16, Vec<u8>)>,
+    url_list: Vec<ServoUrl>,
+    expires: Arc<Mutex<Duration>>,
```

Comment:
Why does this field need its own mutex?

Reply:
You're right that's a mistake, I think I did as a kneejerk reaction when being told by the compiler that the field wasn't mutable, instead of assigning it to a mutable let.